cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java

```diff
@@ -0,0 +1,410 @@
+        new String(ByteStreams.toByteArray(reportIdDir.append(START_FILE).getInputStream()), Charsets.UTF_8);
+      responder.sendJson(new ReportGenerationInfo(creationTime, getReportStatus(reportIdDir), reportRequest));
+    }
+
+    @GET
+    @Path("reports/{report-id}/details")
+    public void getReportDetails(HttpServiceRequest request, HttpServiceResponder responder,
+                                 @PathParam("report-id") String reportId,
+                                 @QueryParam("offset") @DefaultValue("0") long offset,
+                                 @QueryParam("limit") @DefaultValue(DEFAULT_LIMIT) int limit,
+                                 @QueryParam("share-id") String shareId) throws IOException {
+      if (offset < 0) {
+        responder.sendError(400, "offset cannot be negative");
+        return;
+      }
+      if (limit <= 0) {
+        responder.sendError(400, "limit must be a positive integer");
+        return;
+      }
+      if (limit > MAX_LIMIT) {
+        responder.sendError(400, "limit must cannot be larger than " + MAX_LIMIT);
+        return;
+      }
+      Location reportIdDir = getDatasetBaseLocation(ReportGenerationApp.REPORT_FILESET).append(reportId);
+      if (!reportIdDir.exists()) {
+        responder.sendError(404, String.format("Report with id %s does not exist.", reportId));
+        return;
+      }
+      // Get the status of the report and only COMPLETED report can be read
+      ReportStatus status = getReportStatus(reportIdDir);
+      switch (status) {
+        case RUNNING:
+          responder.sendError(202, String.format("Report %s is still being generated, please retry later.", reportId));
+          return;
+        case FAILED:
+          responder.sendError(400, String.format("Reading details of the report %s with failed status is not allowed.",
+                                                 reportId));
+          return;
+        case COMPLETED:
+          break;
+        default: // this should never happen
+          responder.sendError(500, String.format("Unable to read report %s with unknown status %s.", reportId, status));
+          return;
+      }
+      List<String> reportRecords = new ArrayList<>();
+      long lineCount = 0;
+      Location reportDir = reportIdDir.append(REPORT_DIR);
+      Location reportFile = null;
+      // TODO: assume only one report file for now
+      for (Location file : reportDir.list()) {
+        if (file.getName().endsWith(".json")) {
+          reportFile = file;
+          break;
+        }
+      }
+      if (reportFile == null) {
+        responder.sendError(500, "No files found for report " + reportId);
+        return;
+      }
+      // TODO: use cache to store content of the reports
+      // Read the report file and add lines starting from the position of offset to the result until the result reaches
+      // the limit
+      try (BufferedReader br = new BufferedReader(new InputStreamReader(reportFile.getInputStream()))) {
```

Comment:
When creating `InputStreamReader` from `InputStream`, always provide the `Charset`, otherwise you'll run into strange issue on different platform.